thirdparty/s3.go

```diff
@@ -179,7 +189,187 @@ func PutS3File(pushAuth *aws.Auth, localFilePath, s3URL, contentType, permission
+		return legacyPutS3File(pushAuth, localFilePath, s3URL, contentType, permissionACL)
+	}
+
+	urlParsed, err := url.Parse(s3URL)
+	if err != nil {
+		return err
+	}
+
+	if urlParsed.Scheme != "s3" {
+		return errors.Errorf("Don't know how to use URL with scheme %v", urlParsed.Scheme)
+	}
+
+	config := &awsSDK.Config{
+		Credentials: credentials.NewStaticCredentials(pushAuth.AccessKey, pushAuth.SecretKey, pushAuth.Token()),
+		Region:      awsSDK.String(region),
+	}
+	session, err := session.NewSession(config)
+	if err != nil {
+		return errors.Wrap(err, "error creating new session")
+	}
+	svc := awsS3.New(session)
+	bucket := awsS3.CreateBucketInput{
+		Bucket: &urlParsed.Host,
+	}
+
+	file, err := os.Open(localFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "Error opening file %s", localFilePath)
+	}
+	defer file.Close()
+
+	// Get size of file to read into buffer
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return errors.Wrapf(err, "Error getting stats for file %s", localFilePath)
+	}
+	size := fileInfo.Size()
+	buffer := make([]byte, size)
+	_, err = file.Read(buffer)
+	if err != nil {
+		return errors.Wrapf(err, "Error reading bytes of file %S", localFilePath)
+	}
+
+	// Step 1: initiate multipart upload
+	resp, err := createPart(svc, bucket, localFilePath, contentType)
+	if err != nil {
+		return errors.Wrap(err, "Error initiating part")
+	}
+
+	// Step 2: upload each part
+	completedParts := make([]*awsS3.CompletedPart, 0)
+	partNum := int64(1)
+	remaining := size
+	var partLength int64
+	// Keep uploading until there are no many bytes remaining
+	for i := int64(0); remaining != 0; i += partLength {
+		// Abide by 5MB per part limit
+		if remaining < maxPartSize {
+			partLength = remaining
+		} else {
+			partLength = maxPartSize
+		}
+
+		uploadedPart, err := uploadPart(svc, resp, buffer[i:i+partLength], partNum)
+		if err != nil {
+			//  If an error occurs, abort upload to not get charged by Amazon
+			abortErr := abortMultipartUpload(svc, resp)
+			if abortErr != nil {
+				return errors.Wrapf(abortErr, "Error aborting multipart upload for %s", localFilePath)
+			}
+			return errors.Wrapf(err, "Error uploading part in S3 multipart upload for %s", localFilePath)
+		}
+		remaining -= partLength
+		partNum++
+		completedParts = append(completedParts, uploadedPart)
+	}
+
+	// Step 3: complete upload by assembling all uploaded parts
+	return completeMultipartUpload(svc, resp, completedParts)
+}
+
+// createPart initiates a multipart upload
+func createPart(svc *awsS3.S3, bucket awsS3.CreateBucketInput, localFilePath,
+	contentType string) (*awsS3.CreateMultipartUploadOutput, error) {
+	creationInput := &awsS3.CreateMultipartUploadInput{
+		Bucket:      awsSDK.String(*bucket.Bucket),
+		Key:         awsSDK.String(localFilePath),
+		ContentType: awsSDK.String(contentType),
+	}
+
+	result, err := svc.CreateMultipartUpload(creationInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
```

Comment:
What's the purpose of this error handling? It seems like you just return the error

Reply:
You're right, I was following their docs with this error handling, but it looks like they were just doing this for ease of printing. I'm going to remove this and just return the error like usual.